crates/ra_assists/src/merge_match_arms.rs

```diff
@@ -0,0 +1,189 @@
+                        ).collect();
+                    ps.join(" | ")
+                };
+
+                let arm = format!("{} => {}", pats, p_text);
+
+                edit.delete(p.syntax().text_range());
+                // TODO: need to remove trailing comma and whitespace from after removed match
+                edit.delete(n.syntax().text_range());
+                edit.insert(comma.text_range().start(), arm);
+            });
+        }
+    }
+
+    ctx.build()
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::helpers::{check_assist, check_assist_not_applicable};
+    use super::merge_match_arms;
+
+    #[test]
+    fn merge_match_arms_single_patterns() {
+        check_assist(
+            merge_match_arms,
+            r#"
+            #[derive(Debug)]
+            enum X { A, B, C }
+
+            fn main() {
+                let x = X::A;
+                let y = match x {
+                    X::A => { 1i32 },<|>
+                    X::B => { 1i32 },
+                    X::C => { 2i32 }
+                }
+            }
+            "#,
+            r#"
+            #[derive(Debug)]
+            enum X { A, B, C }
+
+            fn main() {
+                let x = X::A;
+                let y = match x {
+                    X::A | X::B => { 1i32 },<|>
+                    X::C => { 2i32 }
+                }
+            }
+            "#
+        );
+    }
+
+    #[test]
+    fn merge_match_arms_multiple_patterns() {
+        check_assist(
+            merge_match_arms,
+            r#"
+            #[derive(Debug)]
+            enum X { A, B, C, D, E }
+
+            fn main() {
+                let x = X::A;
+                let y = match x {
+                    X::A | X::B => { 1i32 },<|>
+                    X::C | X::D => { 1i32 },
+                    X::E => { 2i32 }
+                }
+            }
+            "#,
+            r#"
+            #[derive(Debug)]
+            enum X { A, B, C, D, E }
+
+            fn main() {
+                let x = X::A;
+                let y = match x {
+                    X::A | X::B | X::C | X::D => { 1i32 },<|>
+                    X::E => { 2i32 }
+                }
+            }
+            "#
+        );
+    }
+
+    #[test]
+    fn merge_match_arms_placeholder_pattern() {
+        check_assist(
+            merge_match_arms,
+            r#"
+            #[derive(Debug)]
+            enum X { A, B, C, D, E }
+
+            fn main() {
+                let x = X::A;
+                let y = match x {
+                    X::A => { 1i32 },
+                    X::B => { 2i32 },<|>
+                    _ => { 2i32 }
+                }
+            }
+            "#,
+            r#"
+            #[derive(Debug)]
+            enum X { A, B, C, D, E }
+
+            fn main() {
+                let x = X::A;
+                let y = match x {
+                    X::A => { 1i32 },
+                    _ => { 2i32 },<|>
```

Comment:
Relevant to my question above: I want to make sure that after merging the match expression gets a trailing comma, even if the second arm doesn't initially have one.

Reply:
I think, if the second arm doesn't have a comma, we just shouldn't add one